system/jlib/jptree.cpp

```diff
@@ -766,6 +1110,113 @@ size32_t CPTValue::queryValueSize() const
 
 ///////////////////
 
+
+CPTArray::~CPTArray()
+{
+    if (map.load())
+        map.load();
+}
+
+CQualifierMap *CPTArray::setMap(CQualifierMap *_map)
+{
+    CQualifierMap *expected = nullptr;
+    if (map.compare_exchange_strong(expected, _map))
+        return nullptr;
+    else
+        return map.load();
+}
+
+void CPTArray::addElement(IPropertyTree *tree)
+{
+    append(*tree);
+    CQualifierMap *map = queryMap();
+    if (map)
+    {
+        if (tree->getAttributeCount())
+        {
+            for (auto &lhs: map->getLhsProps())
```

Comment:
SOme of this code should probably move so it is a member function of map

    map->insertAllEntries(tree);

Reply:
yes, refactored.